lib/Dialect/SV/Transforms/SVExtractTestCode.cpp

```diff
@@ -269,6 +272,72 @@ static void migrateOps(hw::HWModuleOp oldMod, hw::HWModuleOp newMod,
     }
 }
 
+// Check if the module has already been bound.
+static bool isBound(hw::HWModuleLike op, hw::InstanceGraph &instanceGraph) {
+  auto *node = instanceGraph.lookup(op);
+  return llvm::any_of(node->uses(), [](hw::InstanceRecord *a) {
+    auto inst = a->getInstance();
+    if (!inst)
+      return false;
+    return inst->hasAttr("doNotPrint");
+  });
+}
+
+// Move any old modules that are test code only to the test code area.
+static void maybeMoveToTestCode(hw::HWModuleOp oldMod, Attribute testBenchDir,
+                                Attribute bindFile,
+                                hw::InstanceGraph &instanceGraph) {
+  // Ensure we have a valid test code path.
+  if (!testBenchDir)
+    return;
+
+  // Check if the module only has inputs.
+  if (oldMod.getNumOutputs() != 0)
+    return;
+
+  // Check if this module has been bound already, and return early if so. This
+  // can happen, for example, if a module is input only, but exists to feed both
+  // asserts and covers. In such cases, the binds end up in the bind file for
+  // the first kind of test code statement we saw, and we expect this to not
+  // cause issues.
+  if (isBound(oldMod, instanceGraph)) {
+    LLVM_DEBUG(oldMod->emitWarning("already bound, skipping for ") << bindFile);
+    return;
+  }
+
+  // Move the module to the test code path.
+  oldMod->setAttr("output_file", testBenchDir);
+
+  // Replace its instances with binds.
+  hw::InstanceGraphNode *node = instanceGraph.lookup(oldMod);
+  OpBuilder b = OpBuilder(oldMod.getContext());
+  for (hw::InstanceRecord *use : node->uses()) {
+    OpBuilder::InsertionGuard g(b);
+
+    hw::HWInstanceLike instLike = use->getInstance();
+    if (!instLike)
+      continue;
+
+    // Mark the instance to not be printed.
+    hw::InstanceOp inst = cast<hw::InstanceOp>(instLike.getOperation());
+    hw::HWModuleLike parent = use->getParent()->getModule();
+    inst->setAttr("doNotPrint", b.getBoolAttr(true));
+
+    // Give the instance an inner sym if it needs one.
+    if (!inst.getInnerSym().has_value())
+      inst.setInnerSymAttr(
+          b.getStringAttr("__" + inst.getInstanceName() + "__"));
+
+    // Create the bind.
+    b.setInsertionPointToEnd(
```

Comment:
The way this currently works, if a module A has some test code in it, and gets instantiated in module Foo and Bar, we will get a bindfile like:

```
bind A A_assert A_assert (...);
bind Foo A a (...);
bind Bar A b (...);
```

I think this makes sense, but if it mattered, we could swap it.

Reply:
I was worried about stability here. However, ETC is running on an (mlir) `ModuleOp`. This should be fine.